### PR TITLE
[WIP] turing TC support [run_process_replay]

### PR DESCRIPTION
**Pull request by ignaciosica**
#### [Reference](https://docs.nvidia.com/cuda/parallel-thread-execution/#warp-level-matrix-fragment-mma-1688)

This PR:
- add Turing TC support for CUDA, PTX and also ops_python
- generalizes CUDA wmma rendering for arbitrary shapes
- ~PTXRenderer now gets `global_max`, `local_max`, `shared_max` and `TCs` from CUDARenderer~ moved to #6836 
- add emulated tests for Turing TC in CI

Generated kernel for matmul with dimensions n8m16k8
``` c
#define INFINITY (__int_as_float(0x7f800000))
#define NAN (__int_as_float(0x7fffffff))
#include <cuda_fp16.h>
struct __align__(8) half4 { half x, y, z, w; }; __device__ half4 make_half4(half x, half y, half z, half w) { half4 r={x, y, z, w}; return r; }
struct __align__(16) half8 { half x, y, z, w, a, b, c, d; }; __device__ half8 make_half8(half x, half y, half z, half w, half a, half b, half c, half d) { half8 r={x, y, z, w, a, b, c, d}; return r; }
__device__ float4 __WMMA_8_16_8_half_float(half4 a, half2 b, float4 c){
  int *a_pk = (int *)(&a), *b_pk = (int *)(&b);
  asm("mma.sync.aligned.m16n8k8.row.col.f32.f16.f16.f32 {%0, %1, %2, %3}, {%4, %5}, {%6}, {%0, %1, %2, %3};"
    : "+f"(c.x), "+f"(c.y), "+f"(c.z), "+f"(c.w)
    : "r"(a_pk[0]), "r"(a_pk[1]), "r"(b_pk[0]));
  return c;
}
extern "C" __global__ void __launch_bounds__(32) r_2_2_2_2_2_4_2_2_2(half* data0, const half* data1, const half* data2) {
  int lidx0 = threadIdx.x; /* 8 */
  int lidx1 = threadIdx.y; /* 2 */
  int lidx2 = threadIdx.z; /* 2 */
  int alu0 = (lidx1*32);
  int alu1 = (lidx2*64);
  int alu2 = (lidx0/4);
  int alu3 = (alu2*16);
  int alu4 = (lidx0%2);
  int alu5 = ((lidx0/2)%2);
  int alu6 = (alu4+(alu5*2)+alu3+alu0+alu1);
  int alu7 = ((alu4*2)+(alu5*4)+alu3+alu0+alu1);
  int alu8 = ((alu4*8)+(alu5*16)+alu2+(lidx1*2)+(lidx2*4));
  half val0 = data1[alu6+4];
  half val1 = data1[alu6+8];
  half val2 = data1[alu6+12];
  half val3 = data1[alu6];
  half val4 = data2[alu8+32];
  half val5 = data2[alu8];
  float4 wmma0 = __WMMA_8_16_8_half_float(make_half4(val3,val0,val1,val2), make_half2(val5,val4), make_float4(0.0f,0.0f,0.0f,0.0f));
  *((half2*)(data0+alu7+8)) = make_half2((half)(wmma0.z),(half)(wmma0.w));
  *((half2*)(data0+alu7)) = make_half2((half)(wmma0.x),(half)(wmma0.y));
}
```

### CODE REVIEW

File: tinygrad/renderer/cstyle.py [hunk 1]
```diff
@@ -351,14 +356,16 @@ def render_kernel(self, function_name, kernel, bufs, uops, prefix=None):
...
 
-    # TODO: this has to be way better to generate for arbitrary M,N,K: use arg[1] for MNK, use arg[4] for vec sizes, encode register packing
     dt_map = { dtypes.float: "f32", dtypes.half: "f16", dtypes.bfloat16: "bf16" }
-    for arg in dedup([uop.arg for uop in uops if uop.op is UOps.WMMA]):
-      fn, ti, to, ci, co = arg[0], self.render_dtype(arg[2]), self.render_dtype(arg[3]), dt_map[arg[2]], dt_map[arg[3]]
-      prefix.append(f"""__device__ {to}4 __{fn}({ti}8 a, {ti}4 b, {to}4 c) {{ int *a_pk = (int *) (&a), *b_pk = (int *) (&b);
-asm( "mma.sync.aligned.m16n8k16.row.col.{co}.{ci}.{ci}.{co} {{ %0, %1, %2, %3 }}, {{ %4, %5, %6, %7 }}, {{ %8, %9 }}, {{ %0, %1, %2, %3 }};"
-  : "+f"(c.x), "+f"(c.y), "+f"(c.z), "+f"(c.w) : "r"(a_pk[0]), "r"(a_pk[1]), "r"(a_pk[2]),  "r"(a_pk[3]), "r"(b_pk[0]), "r"(b_pk[1]) );
-return c;}}""")
+    for name, (N, M, K), dti, dto, _, _, (upc_a, upc_b, upc_c), _ in dedup([uop.arg for uop in uops if uop.op is UOps.WMMA]):
```
Comment (geohot) [inline tinygrad/renderer/cstyle.py:360]: This chunk of code is totally unreadable. Feel free to use more lines to make it so I can actually read it

File: tinygrad/renderer/assembly.py [hunk 2]
```diff
@@ -82,15 +83,14 @@ def render_val(x, dtype):
...
-  global_max = (2147483647, 65535, 65535)
-  local_max = (1024, 1024, 64)
-  shared_max = 49152
   tensor_cores = [tc for tc in CUDARenderer.tensor_cores if tc.dtype_in == dtypes.half]
   code_for_op = asm_for_op
   extra_matcher = ptx_matcher
-  def __init__(self, arch:str, device="CUDA"): self.device, self.tensor_cores = device, PTXRenderer.tensor_cores if int(arch[3:]) >= 80 else []
+  def __init__(self, arch:str, device="CUDA"):
+    self.device = device
+    self.global_max, self.local_max, self.shared_max = CUDARenderer.global_max, CUDARenderer.local_max, CUDARenderer.shared_max
```
Comment (geohot) [inline tinygrad/renderer/assembly.py:92]: Err, do these change? Why aren't they global scope?

File: tinygrad/renderer/assembly.py [hunk 3]
```diff
@@ -250,14 +250,15 @@ def _cast(a, dtype:DType, atype:DType, bitcast=False, u=None, pred=False):
           dt = dtypes.ulong if dtype.__class__ == PtrDType else dtype
           kk(*self.render_load(nm, ssa('dat', u, self.types[dt]), dt, ss=".param"))
         elif uop is UOps.WMMA:
-          wmma = []
+          _, (N, M, K), dti, _, _, _, (upc_a, _, _), _ = args
```
Comment (geohot) [inline tinygrad/renderer/assembly.py:253]: dti, upc_a? what is this

Comment (ignaciosica) [reply tinygrad/renderer/assembly.py]: fixed

File: tinygrad/renderer/assembly.py (hunk 2 above)
Comment (ignaciosica) [reply tinygrad/renderer/assembly.py]: They don't, you are right, they should be global scope

File: tinygrad/renderer/cstyle.py (hunk 1 above)
Comment (ignaciosica) [reply tinygrad/renderer/cstyle.py]: refactored, hopefully now is more readable

**Comment by geohot** [comment]
Was this tested by people with Turing?

**Comment by ignaciosica** [comment]
> Was this tested by people with Turing?

nobody tested it yet

**Pull request closed without merging**